dipy/align/vector_fields.pyx

```diff
@@ -3065,3 +3065,411 @@ def create_sphere(cnp.npy_intp nslices, cnp.npy_intp nrows,
+                # Interpolate img at q
+                in_flag = _interpolate_scalar_2d(img, q[0], q[1],
+                                                &out[i, p])
+                if in_flag == 0:
+                    out[i, p] = 0
+                    inside[i] = 0
+                    continue
+                tmp = out[i, p]
+                # Compute coordinates of point dx on img's grid
+                dx[p] = sample_points[i, p] + h[p]
+                q[0] = _apply_affine_2d_x0(dx[0], dx[1], 1, img_affine_inv)
+                q[1] = _apply_affine_2d_x1(dx[0], dx[1], 1, img_affine_inv)
+                # Interpolate img at q
+                in_flag = _interpolate_scalar_2d(img, q[0], q[1],
+                                                &out[i, p])
+                if in_flag == 0:
+                    out[i, p] = 0
+                    inside[i] = 0
+                    continue
+                out[i, p] = (out[i, p] - tmp) / img_spacing[p]
+                dx[p] = sample_points[i, p]
+
+
+def gradient(img, img_affine_inv, img_spacing, domain_shape, domain_affine):
+    r""" Gradient of an image in physical space
+
+    Parameters
+    ----------
+    img : 2D or 3D array, shape (R, C) or (S, R, C)
+        the input image whose gradient will be computed
+    img_affine_inv : array, shape (dim+1, dim+1)
+        the space-to-grid transform matrix associated to img
+    img_spacing : array, shape (dim,)
+        the spacing between voxels (voxel size along each axis) of the input
+        image
+    domain_shape : array, shape (dim,)
+        the number of (slices), rows and columns of the sampling grid
+    domain_affine : array, shape (dim+1, dim+1)
+        the grid-to-space transform associated to the sampling grid
+
+    Returns
+    -------
+    out : array, shape (R', C', 2) or (S', R', C', 3)
+        the buffer in which to store the image gradient, where
+        (S'), R', C' are given by domain_shape
+    """
+    dim = len(img.shape)
+    if not is_valid_affine(img_affine_inv, dim):
+        raise ValueError("Invalid image affine transform")
+    if not is_valid_affine(domain_affine, dim):
+        raise ValueError("Invalid sampling grid affine transform")
+    if len(img_spacing) < dim:
+        raise ValueError("Invalid spacings")
+    ftype = img.dtype
+    out = np.ndarray(tuple(domain_shape)+(dim,), dtype=ftype)
+    inside = np.ndarray(tuple(domain_shape), dtype=np.int32)
+    if dim == 2:
+        _gradient_2d(img, img_affine_inv.astype(np.float64),
+                     img_spacing.astype(np.float64),
+                     domain_affine.astype(np.float64), out, inside)
+    else:
+        _gradient_3d(img, img_affine_inv.astype(np.float64),
+                     img_spacing.astype(np.float64),
+                     domain_affine.astype(np.float64), out, inside)
+    return out, inside
+
+
+def sparse_gradient(img, img_affine_inv, img_spacing, sample_points):
+    r""" Gradient of an image in physical space
+
```

Comment:
I wonder if we should expose those these functions beyond registration as they can be used for example for segmentation. We should think if in the future we should have a tutorial exposing those functions that interpolate and calculate gradients on 3D images. It could be useful for other projects. Think of it.